### DISCUSSION TITLE
The Arbitrary Variants release

### DISCUSSION THREAD

**Discussion by ben-rogerson** [ben-rogerson/twin.macro#556] [Releases]
### 🌟  New: Arbitrary _Variants_ (#546)

**The star feature of this release is the addition of a new variant that creates a custom selector for classes**.

Usage is easy, just add the selector within square brackets and prefix it on any class (or round-bracketed set):

```js
// Style child elements with the strong tag from the parent container
;<div tw="[strong]:text-red-500">
    <strong>I'm red</strong>
</div>
```

Before this feature, the styling above would need help from the css prop:

```js
// Style child elements with the strong tag from the parent container
;<div css={{ strong: tw`text-red-500` }}>
    <strong>I'm red</strong>
</div>
```

There are a wide range of uses for Arbitrary Variants:

```js
// Style the current element based on a dynamic className
;<div className={isSmall && 'is-small'} tw="text-lg [&.is-small]:text-sm">...</div>

// Style the current element based on a theming/scoping className
;<body className="dark-theme">
    <div tw="[.dark-theme &]:(bg-black text-white)">I'm dark theme</div>
</body>

// Add custom group selectors
;<button className="group" disabled>
    <span tw="[.group:disabled &]:text-gray-500">I'm gray</span>
</button>

// Use custom at-rules like @supports
;<div tw="[@supports (display: grid)]:grid">I'm grid</div>

// Or custom height queries
;<div tw="[@media (min-height: 800px)]:hidden">
    I'm shown only on smaller window heights
</div>
```

I hope you enjoy using this powerful feature!

### 🐛  Bugfixes

- Allow dot notation components to be styled with twin (eg: `<Menu.Item tw="block">`) (#534)
- Add array value types for the screen import (#555) - thanks @mutewinter
- Avoid `bg-color` classes from trumping `bg-opacity-x` classes (#552)
- Add helpful feedback when a `peer` class is used as a tailwind class (#551)
- Convert all numeric values to strings (#540)
- Fix issue with css variables declared in `:root` using `addBase` in plugins (#521)
- Enable variants to be used with the `container` class (#518)
- Restrict debug mode to development (#497) - thanks @u3u

<hr /><em>This discussion was created from the release <a href='https://github.com/ben-rogerson/twin.macro/releases/tag/2.8.0'>The Arbitrary Variants release</a>.</em>

**Comment by jvanderen1** [comment]
Amazing! This can let us now style input components using the shadow dom or even nested react components with needing to pass complex props 🤯

Such a great addition and powerful feature! 👏🏻
